File: frontend/src/services/notification.service.ts
import api from './api';
import type { Notification } from '../types';

const BASE = '/notifications';

export const notificationService = {
  /** Fetch the most recent N notifications (default 20) */
  getRecent: async (limit = 20): Promise<Notification[]> => {
    const { data } = await api.get<Notification[]>(`${BASE}/recent`, { params: { limit } });
    return data;
  },

  /** Fetch only unread notifications */
  getUnread: async (): Promise<Notification[]> => {
    const { data } = await api.get<Notification[]>(`${BASE}/unread`);
    return data;
  },

  /** Get the count of unread notifications */
  getUnreadCount: async (): Promise<number> => {
    const { data } = await api.get<number>(`${BASE}/count`);
    return data;
  },

  /** Mark a single notification as read */
  markAsRead: async (id: number): Promise<void> => {
    await api.patch(`${BASE}/${id}/read`);
  },

  /** Mark all notifications as read */
  markAllAsRead: async (): Promise<void> => {
    await api.patch(`${BASE}/read-all`);
  },
};

File: backend/src/main/java/com/pgsi/service/NotificationService.java
package com.pgsi.service;

import com.pgsi.dto.NotificationDto;

import java.util.List;

public interface NotificationService {

    /**
     * Create and persist an in-app notification for a single recipient.
     */
    void createNotification(Long recipientId, String type, String title, String message, Long relatedTicketId);

    /**
     * Return all unread notifications for the current authenticated user.
     */
    List<NotificationDto> getUnreadNotifications(String username);

    /**
     * Return the last N notifications (read + unread) for the current user.
     */
    List<NotificationDto> getRecentNotifications(String username, int limit);

    /**
     * Count unread notifications for the current user.
     */
    long countUnread(String username);

    /**
     * Mark a single notification as read.
     */
    void markAsRead(Long notificationId, String username);

    /**
     * Mark all notifications as read for the current user.
     */
    void markAllAsRead(String username);
}
